addons/microsoft_calendar/models/microsoft_sync.py

```diff
@@ -131,6 +131,8 @@ def _sync_odoo2microsoft(self, microsoft_service: MicrosoftCalendarService):
             records_to_sync = self
         cancelled_records = self - records_to_sync
 
+        records_to_sync._ensure_attendees_have_email()
+        records_to_sync = records_to_sync.with_context(_attendees_have_email=True)
```

Comment:
Why do you call the _ensure_attendees_have_email method here and once again in _microsoft_patch and _ensure_attendees_have_email ? I don't understand why do you need the context key _attendees_have_email.
Don't you need to check mails everytime ? When you insert/patch without coming from _sync_odoo2microsoft, you don't need to check that attendees have mails? The context key will be missing. So if you only need to check that from _sync_odoo2microsoft, don't add self._ensure_attendees_have_email() in match and insert.

Reply:
My problem is that there are two possibilities:
1. Suppose the user has 100 incorrect events and the calendars are syncing. Since we want to provide a complete report to the user, we check all events in `_sync_odoo2microsoft`. However, if all events are correct, `_microsoft_insert` or `_microsoft_patch` is called for each event, so we need to prevent `_ensure_attendees_have_email` from running again.
2. The calendars are synced and the user adds/updates an existing event. This directly calls `_microsoft_insert`/`_microsoft_patch`. Since `_sync_odoo2microsoft` is not called, we need to `_ensure_attendees_have_email`. Note that, in this case, since the `ValidationError` is caught, this will only prevent the event to be created/updated on Microsoft side. Thereafter, `_sync_odoo2microsoft` will eventually be called and the user will know that an event has to be updated. (If we don't prevent the event to be created/updated on Microsoft side, we fall back into the initial problem of this RP)

So, `_attendees_have_email` allows us to differentiate between the two situations.